add mlkem768x25519-sha256 key exchange support

lucasqueiroz23 opened the pull request:
This implements mlkem768x25519-sha256 according to the following draft: https://datatracker.ietf.org/doc/html/draft-ietf-sshm-mlkem-hybrid-kex-03.html#name-mlkem768x25519-sha256

This is a Post-Quantum/Traditional hybrid key exchange algorithm. Basically, this means that we use both a traditional key exchange mechanism and a post-quantum one to generate a shared secret. 

Post-Quantum algorithms are those that should not break in face of a quantum computer attack. This one is hybrid because even if mlkem proves to break from a traditional computer attack, curve25519 will serve as a fallback. 

Also, OpenSSH v10.0+ uses this same algorithm as default (https://www.openssh.org/pq.html). 

This implementation is limited to node 24.7.0+, because that's when node got mlkem support on its crypto API.

Relevant document:
https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.203.pdf

lucasqueiroz23 committed:
feat: checks if mlkem is supported and lints files

lucasqueiroz23 committed:
feat: adds mlkem class, adds it to factory and adds generateKeys()

lucasqueiroz23 committed:
feat: adds getPublicKey() and start()

lucasqueiroz23 committed:
fix: changes key pair function used to make it sync

lucasqueiroz23 committed:
feat: adds more context to support check

lucasqueiroz23 committed:
fix: changes constant name to correct one

lucasqueiroz23 committed:
feat: removes async logic and finishes implementation

lucasqueiroz23 committed:
docs: adds mlkem768x25519-sha256 as kex option

mscdex commented:
All of the unrelated (code style) changes need to be reverted. Additionally for the new changes, make sure it at least passes the linting (`npm run lint`).

mscdex commented:
We also need at least one test that connects the client and server using the new algorithm to at least make sure that works.

Unfortunately OpenSSH 9.6 is currently the latest version on Gitub Actions' `ubuntu-latest`, so there is no way to test with the OpenSSH client in CI.

lucasqueiroz23 committed:
style: reverts style changes

lucasqueiroz23 commented:
> All of the unrelated (code style) changes need to be reverted. Additionally for the new changes, make sure it at least passes the linting (`npm run lint`).

Done! Didn't notice that, my text editor made these changes automatically. 



> We also need at least one test that connects the client and server using the new algorithm to at least make sure that works.
> 
> Unfortunately OpenSSH 9.6 is currently the latest version on Gitub Actions' `ubuntu-latest`, so there is no way to test with the OpenSSH client in CI.

Ok, I'll be working on it tomorrow.

lucasqueiroz23 committed:
fix: hashes key concatenation

lucasqueiroz23 committed:
test: adds integration tests for mlkem

lucasqueiroz23 commented:
@mscdex In order to test with OpenSSH, I created a branch on my fork (https://github.com/lucasqueiroz23/ssh2/tree/mlkem-tests) that creates two docker containers: one that'll run an OpenSSH10.1 server and another that'll run some scripts that use ssh2.js client and server implementations. Since I won't open a PR for that branch, I updated its README to describe its usage. I dockerized it so you could easily test it as well, if you want to.

Basically, we'll have a script that'll run a ssh2.js client (which accepts mlkem as kex) that connects to the openssh10.1 server, another that runs a ssh2.js server (that accepts mlkem as kex) and connects an openssh10.1 client to it and an ssh2.js server and a client that will both accept mlkem and connect to each other.  

It took some time, but I think it was worth it. 

Here's the output:
<img width="827" height="697" alt="image" src="https://github.com/user-attachments/assets/adbab1e4-acee-431b-a970-9e94634e57f0" />

If I add the debug feature on each test, we can see that the handshake works fine for each of them:

<img width="567" height="843" alt="image" src="https://github.com/user-attachments/assets/1ba0ae44-0e77-4bbc-9b58-9aa600b4f8f3" />

<img width="704" height="815" alt="image" src="https://github.com/user-attachments/assets/7093869a-dc2a-454b-a848-d8def8a9990b" />

<img width="704" height="761" alt="image" src="https://github.com/user-attachments/assets/ce20ef85-1499-4b34-979f-b52f907d0fb6" />


I also added some integration tests on `test/test-misc-client-server.js`. I think the most interesting one checks if client and server computed the same `K_PQ` and `K_CL` values.

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+      this._curve25519KeySize = 32;
+      this._encapsulationKeySize = 1184;
+      this._decapsulationKeySize = 2400;
+      this._ciphertextSize = 1088;
```
Constants don't need to be added to the object. If we absolutely need them, it's better to just make them actual constants (with names that are short and sweet and to the point) scoped to this file instead.

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+      this._ciphertextSize = 1088;
+
+      this.K_PQ = null;
+      this.K_CL = null;
```
We don't need to add these just for the purposes of testing. Once Github Actions updates their `ubuntu-latest`, we can just rely on OpenSSH compatibility testing to further ensure that everything is working as intended.

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+
+      if (!this._mlkemKeys)
+        this._mlkemKeys = generateKeyPairSync('ml-kem-768');
+
```
```suggestion
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -621,7 +624,6 @@ const createKeyExchange = (() => {
 
       // "H"
       const exchangeHash = hash.digest();
-
```
This needs to be reverted.

mscdex commented on lib/protocol/constants.js:
```diff
@@ -32,6 +32,38 @@ const curve25519Supported = (typeof crypto.diffieHellman === 'function'
+      || typeof crypto.encapsulate !== 'function'
+      || typeof crypto.decapsulate !== 'function'
+    )
+      return false;
```
```suggestion
    if (!curve25519Supported
        || typeof crypto.encapsulate !== 'function'
        || typeof crypto.decapsulate !== 'function') {
      return false;
    }
```

mscdex commented on lib/protocol/constants.js:
```diff
@@ -32,6 +32,38 @@ const curve25519Supported = (typeof crypto.diffieHellman === 'function'
+      return false;
+
+    const algorithm = 'ml-kem-768';
+    const test_clientKeys = crypto.generateKeyPairSync(algorithm);
```
```suggestion
    const keys = crypto.generateKeyPairSync('ml-kem-768');
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+      key.copy(asnWriter._buf,
+        asnWriter._offset,
+        0,
+        key.length);
```
```suggestion
      key.copy(
        asnWriter._buf,
        asnWriter._offset,
        0,
        key.length
      );
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+        .update(K_CL)
+        .digest();
+      
+        return K;
+
+
```
```suggestion
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+      } catch (error) {
+        return error;
+      }
+
```
```suggestion
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1582,230 @@ const createKeyExchange = (() => {
+          const K = createHash('sha256')
+          .update(K_PQ)
+          .update(K_CL)
+          .digest();
```
```suggestion
            .update(K_PQ)
            .update(K_CL)
            .digest();
```

mscdex commented on lib/protocol/constants.js:
```diff
@@ -32,6 +32,38 @@ const curve25519Supported = (typeof crypto.diffieHellman === 'function'
+    for (let i = 0; i < sharedSecret.length && i < sharedKey.length; i++) {
+      if (sharedSecret[i] !== sharedKey[i])
+        return false;
+    }
```
I don't think we need to do this. As long as the node APIs exist and no error is thrown, we can trust node to do the right thing.

lucasqueiroz23 committed:
refactor: applies suggestions

lucasqueiroz23 commented:
@mscdex Suggestions applied!

mscdex commented on lib/protocol/kex.js:
```diff
@@ -9,13 +9,16 @@ const {
   generateKeyPairSync,
   randomFillSync,
+  encapsulate,
+  decapsulate,
```
These should remain in alphabetic order (for the entire destructuring).

mscdex commented on lib/protocol/kex.js:
```diff
@@ -9,13 +9,16 @@ const {
 const {
   COMPAT,
   curve25519Supported,
+  mlkemSupported,
```
This should remain in alphabetic order (for the entire destructuring).

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1583,210 @@ const createKeyExchange = (() => {
+          const C_PK2 = otherPublicKey.slice(
+            0,
+            1184
+          );
```
```suggestion
          const C_PK2 = otherPublicKey.slice(0, 1184);
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1583,210 @@ const createKeyExchange = (() => {
+          C_PK2.copy(
+            C_PK2_DER,
+            templateDER.length - 1184
+          );
```
```suggestion
          templateDER.copy(C_PK2_DER, 0, 0, templateDER.length - 1184);

          C_PK2.copy(C_PK2_DER, templateDER.length - 1184);
```

mscdex commented on lib/protocol/kex.js:
```diff
@@ -1580,13 +1583,210 @@ const createKeyExchange = (() => {
+        const K_PQ = decapsulate(
+          this._mlkemKeys.privateKey,
+          S_CT2
+        );
```
```suggestion
        const K_PQ = decapsulate(this._mlkemKeys.privateKey, S_CT2);
```

lucasqueiroz23 committed:
refactor: apply suggestions from code review

Co-authored-by: mscdex <mscdex@users.noreply.github.com>

lucasqueiroz23 committed:
refactor: imports in alphabetic order
